samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -21,99 +21,71 @@ package org.apache.samza.system.kafka
 
 
 import java.util.concurrent.atomic.AtomicReference
-import java.util.concurrent.{TimeUnit, ConcurrentHashMap, Future}
+import java.util.concurrent.TimeUnit
 
+import kafka.producer.ProducerClosedException
 import org.apache.kafka.clients.producer.Callback
 import org.apache.kafka.clients.producer.Producer
 import org.apache.kafka.clients.producer.ProducerRecord
 import org.apache.kafka.clients.producer.RecordMetadata
 import org.apache.kafka.common.PartitionInfo
-import org.apache.samza.SamzaException
+import org.apache.kafka.common.errors.SerializationException
 import org.apache.samza.system.OutgoingMessageEnvelope
 import org.apache.samza.system.SystemProducer
+import org.apache.samza.system.SystemProducerException
 import org.apache.samza.util.ExponentialSleepStrategy
 import org.apache.samza.util.KafkaUtil
 import org.apache.samza.util.Logging
 import org.apache.samza.util.TimerUtils
 
-import scala.collection.JavaConverters._
+import scala.collection.mutable
 
 class KafkaSystemProducer(systemName: String,
                           retryBackoff: ExponentialSleepStrategy = new ExponentialSleepStrategy,
                           getProducer: () => Producer[Array[Byte], Array[Byte]],
                           metrics: KafkaSystemProducerMetrics,
-                          val clock: () => Long = () => System.nanoTime) extends SystemProducer with Logging with TimerUtils
+                          val clock: () => Long = () => System.nanoTime,
+                          val dropProducerExceptions: Boolean = false) extends SystemProducer with Logging with TimerUtils
 {
 
-  class SourceData {
-    /**
-     * lock to make send() and store its future atomic
-     */
-    val sendLock: Object = new Object
-    /**
-     * The most recent send's Future handle
-     */
-    @volatile
-    var latestFuture: Future[RecordMetadata] = null
-    /**
-     * exceptionInCallback: to store the exception in case of any "ultimate" send failure (ie. failure
-     * after exhausting max_retries in Kafka producer) in the I/O thread, we do not continue to queue up more send
-     * requests from the samza thread. It helps the samza thread identify if the failure happened in I/O thread or not.
-     *
-     * In cases of multiple exceptions in the callbacks, we keep the first one before throwing.
-     */
-    var exceptionInCallback: AtomicReference[SamzaException] = new AtomicReference[SamzaException]()
-  }
-
-  @volatile var producer: Producer[Array[Byte], Array[Byte]] = null
+  /**
+    * Represents a fatal error that caused the producer to close.
+    */
+  val fatalException: AtomicReference[SystemProducerException] = new AtomicReference[SystemProducerException]()
+  var producer: Producer[Array[Byte], Array[Byte]] = null
```

Comment:
IIUC this needs to be volatile (or its reads need to be synchronized on producerLock too). Otherwise sender/flushing threads aren't guaranteed to see the new producer created by the Kafka I/O thread in the callback (i.e., there's no memory barrier).